### PR TITLE
Add formatted output  - #93

### PR DESCRIPTION
**Pull request by kvig**
This adds CSV and JSON formatted output for Functionary results.

The `package.yaml` schema was updated to add the `return_type` and `output_format` parameters.
Supported values for `return_type` are:
- string
- json
- csv
Supported values for `output_format` are:
- json
- table
The task details page will do its best to coerce strings and json `result_type`'s to a JSON object.

In the case of an `output_format` set to `table`, csv and string `result_type`'s are assumed to be a string in csv format. These are parsed with the csv lib and the results set for rendering. For a json `result_type`, the output is assumed to be a JSON list or a JSON object with one key to a list of values, for example:
```json
[
  {
    "prop1":"value1",
    "prop2":"value2"
  },
  {
    "prop1":"value3",
    "prop2":"value4"
  }
]
```
or
```json
{
  "foos" : [
    {
      "prop1":"value1",
      "prop2":"value2"
    },
    {
      "prop1":"value3",
      "prop2":"value4"
    }
  ]
}
```

In all cases, if the conversion cannot be determined or there is an exception raised, the output will be the raw result and a message logged on the server.

### CODE REVIEW

**Review by scott-taubman — COMMENTED** [review]
I made a few comments, but seeing the way this shaping out it makes me want to take a step back and rethink somewhat the target that we're after here.  More discussion to come.

File: functionary/ui/templates/tags/output_table.html [hunk 1]
```diff
@@ -0,0 +1,18 @@
+{% if format == "table" %}
+    <table id="result_table" class="result_table">
```
[the remark this answers is not in the document]
Comment (kvig) [reply functionary/ui/templates/tags/output_table.html:2]: Good call, completely forgot about Bulma

File: functionary/builder/utils.py [hunk 2]
```diff
@@ -227,13 +238,19 @@ def _create_functions_from_definition(definitions, package: Package):
     db_functions = []
     for function_def in definitions:
         name = function_def.get("name")
+        ret_type = function_def.get("return_type")
+        if ret_type not in type_map:
+            raise ValueError(f"Incorrect return type for {name}")
```
[the remark this answers is not in the document]
Comment (kvig) [reply functionary/builder/utils.py:243]: Yes, that's definitely where it should be at

File: functionary/ui/templatetags/task_output.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+                        r.append(row[key])
+                    res_data.append(r)
+
+                result = {"headers": headers, "data": res_data}
+            else:
+                raise ValueError("Unable to determine the structure of the JSON")
+        else:
+            raise ValueError("JSON data appears to be empty")
+    else:
+        raise ValueError(f"Unable to convert {result_type} to csv")
```
Comment (scott-taubman) [inline functionary/ui/templatetags/task_output.py:79]: Mostly just a nitpick for the sake of clarity, but both here and in the comment a few lines belong you refer to converting data to CSV.  It doesn't look like that's what you're doing necessarily.  You're organizing the data for purposes of displaying it as a table, right?

File: functionary/ui/templatetags/task_output.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+                    e,
+                )
+
+        return f"<pre>{'<br/>'.join(the_task.result.splitlines())}</pre>"
+
+
+@register.tag("task_output")
+def do_task_output(parser, token):
+    """Convert the result of the task to the specified format."""
+    _, rest = token.contents.split(None, 1)
```
Comment (scott-taubman) [inline functionary/ui/templatetags/task_output.py:133]: I don't know what the right name for `rest`, but it's not obvious what data that refers to.  Is `tag_param` or something similar a better name?

File: functionary/ui/templates/tags/output_table.html [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+                {% for row in result.data %}
+                    <tr>
+                        {% for col in row %}<td>{{ col }}</td>{% endfor %}
+                    </tr>
+                {% endfor %}
+            </tbody>
+        </table>
+    </div>
+{% else %}
+    <pre>{{ result | linebreaksbr }}</pre>
```
Comment (scott-taubman) [inline functionary/ui/templates/tags/output_table.html:19]: Seeing this in both templates strikes me as weird.  Should this just be its own `output_raw.html` template that gets rendered if the output_format is neither of json or html?

File: functionary/ui/templatetags/extras.py [hunk 6]
```diff
@@ -10,7 +10,7 @@ def pretty_json(value):
     return json.dumps(value, indent=4)
 
 
-@register.filter
+@register.filter(is_safe=True)
 def input_type(value):
```
[the remark this answers is not in the document]
Comment (kvig) [reply functionary/ui/templatetags/extras.py:14]: This is used for the parameters when the task form is displayed. Without this, the HTML inputs are essentially character fields allowing anything to be input vs numbers.

File: functionary/ui/templatetags/task_output.py (hunk 4 above)
Comment (kvig) [reply functionary/ui/templatetags/task_output.py:133]: yes, that was a bad name

File: functionary/ui/templates/tags/output_table.html (hunk 5 above)
Comment (kvig) [reply functionary/ui/templates/tags/output_table.html:19]: That was an artifact from not having separate output templates. It's not accessible now so it has been removed.

**Review by scott-taubman — APPROVED** [review]

**Pull request merged**